bpfman-operator/controllers/bpfman-agent/common.go

```diff
@@ -74,19 +74,234 @@ type ReconcilerCommon struct {
-		*bpfmaniov1alpha1.BpfProgram,
-		bool,
-		bool,
-		*MapOwnerParamStatus) (bpfmaniov1alpha1.BpfProgramConditionType, error)
+	// SetupWithManager registers the reconciler with the manager and defines
+	// which events will trigger a reconcile.
+	SetupWithManager(mgr ctrl.Manager) error
+	// Reconcile is the main entry point to the reconciler. It will be called by
+	// the controller runtime when something happens that the reconciler is
+	// interested in. When Reconcile is invoked, it initializes some state in
+	// the given bpfmanReconciler, retrieves a list of all programs of the given
+	// type, and then calls reconcileCommon.
+	Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error)
+	// getFinalizer returns the string used for the finalizer to prevent the
+	// BpfProgram object from deletion until cleanup can be performed
 	getFinalizer() string
+	// getRecType returns the type of the reconciler.  This is often the string
+	// representation of the ProgramType, but in cases where there are multiple
+	// reconcilers for a single ProgramType, it may be different (e.g., uprobe,
+	// fentry, and fexit)
 	getRecType() string
-	expectedBpfPrograms(ctx context.Context) (*bpfmaniov1alpha1.BpfProgramList, error)
+	// getProgType returns the ProgramType used by bpfman for the bpfPrograms
+	// the reconciler manages.
+	getProgType() internal.ProgramType
+	// getName returns the name of the current program being reconciled.
+	getName() string
+	// getExpectedBpfPrograms returns the list of BpfPrograms that are expected
+	// to be loaded on the current node.
+	getExpectedBpfPrograms(ctx context.Context) (*bpfmaniov1alpha1.BpfProgramList, error)
+	// getLoadRequest returns the LoadRequest that should be sent to bpfman to
+	// load the given BpfProgram.
+	getLoadRequest(bpfProgram *bpfmaniov1alpha1.BpfProgram, mapOwnerId *uint32) (*gobpfman.LoadRequest, error)
+	// getNode returns node object for the current node.
+	getNode() *v1.Node
+	// getBpfProgramCommon returns the BpfProgramCommon object for the current
+	// Program being reconciled.
+	getBpfProgramCommon() *bpfmaniov1alpha1.BpfProgramCommon
+	// setCurrentProgram sets the current program for the reconciler as well as
+	// any other related state needed.
+	setCurrentProgram(program client.Object) error
+}
+
+// reconcileCommon is the common reconciler loop called by each bpfman
+// reconciler.  It reconciles each program in the list.  reconcileCommon should
+// not return error because it will trigger an infinite reconcile loop.
+// Instead, it should report the error to user and retry if specified. For some
+// errors the controller may decide not to retry. Note: This only results in
+// calls to bpfman if we need to change something
+func (r *ReconcilerCommon) reconcileCommon(ctx context.Context, rec bpfmanReconciler,
+	programList []client.Object) (ctrl.Result, error) {
+
+	r.Logger.V(1).Info("Start reconcileCommon(), programList:")
+	for _, program := range programList {
```

Comment:
I'm going to delete the loop. The names get printed out below, but I added this for debugging because it wasn't working the way I expected.  Now it works right, so I don't think it's needed.